### DISCUSSION TITLE
I have two questions about grammar

### DISCUSSION THREAD

**Discussion by Pete-mathpresso** [lark-parser/lark#887] [Q&A]
What should I do if I don't want to ignore some of the grammar in grammar and some of it? In the Lark grammar, %ignore " " is used to control the spaces in all rules and terminal definitions, but I want to control only some of them.
For example, when you create a grammar that takes and parses a multiplication with latex-form of alphabet, integer and decimal numbers do not want spaces on both sides of the point, but I want to ignore them otherwise. All I can think of is using regular expressions. I wonder if there is a module that can be provided by the Lark library, such as the 'Combine' module provided by Pyparsing. (It's 1st Question)
[2nd Question]
I want to implement a grammar that parses integer or decimal characters, but I am having difficulty using or grammar.How do we deal with this?
I wrote like this,
grammar = 
r"""
    expr : integer | decimal_default
    
    integer : INT
    decimal_default : /\d+\.\d+/
    
    %import common.INT
    %import common.WS
    %ignore WS

"""

parser = Lark(grammar, parser='lalr', start='expr')
test_text = r'3.5'
test_result = parser.parse(test_text)
result = parser.parse(test_text)
print('result : ', result)

But there is a bug report like that,
![스크린샷 2021-05-04 오전 11 48 29](https://user-images.githubusercontent.com/60593824/116955856-a7007e80-acce-11eb-8ece-eb0ed2198527.png)

**Comment by erezsh** [comment]
1)  You can simply add spaces to the terminal. Like so:
```
from lark import Lark

p = Lark("""
start: "a" "b" " c"
_WS: " "
%ignore _WS
""", parser="lalr")

p.parse("a b c")	# OK
p.parse("a bc")		# Error
```

If you were using Earley, you could even write `_WS "c"` instead.

2) The lexer is matching `INT` before the decimal. Simply define decimal with a high priority:

```
DECIMAL.2: /\d+\.\d+/
```

**Reply by Pete-mathpresso** [reply]
thank you for answer it,
1. Actually, I want control WS completely like Combine module in Pyparsing. This is docstring about Combine
    """
    Converter to concatenate all matching tokens to a single string.
    By default, the matching patterns must also be contiguous in the input string;
    this can be disabled by specifying C{'adjacent=False'} in the constructor.

    Example::
        real = Word(nums) + '.' + Word(nums)
        print(real.parseString('3.1416')) # -> ['3', '.', '1416']
        # will also erroneously match the following
        print(real.parseString('3. 1416')) # -> ['3', '.', '1416']

        real = Combine(Word(nums) + '.' + Word(nums))
        print(real.parseString('3.1416')) # -> ['3.1416']
        # no match when there are internal spaces
        print(real.parseString('3. 1416')) # -> Exception: Expected W:(0123...)
    """
There are the ones I'm looking for.
  
![스크린샷 2021-05-04 오후 1 12 20](https://user-images.githubusercontent.com/60593824/116960098-660e6700-acda-11eb-8964-3eb8b1c9a487.png)

    
I think It would be essential to ignore WS to automatically remove all large spaces. and In such a situation, a module such as Combine would be needed to treat the absence as an error. If not, the rules of grammar might be dirty and complicated.
   
2. Also same error even I swith the integer and decimal_default rules and give priority to decimal_default
grammar = r"""
    expr : decimal_default | integer
    
    integer : INT
    decimal_default.2 : /\d+\.\d+/
    
    %import common.INT
    %import common.WS
    %ignore WS
"""

test_text = r'3.5'


parser = Lark(grammar, parser='lalr', start='expr')
result = parser.parse(test_text)

print('result : ', result)
print('result_type : ', type(result))

-> same error

**Reply by erezsh** [reply]
1. Accomplishing this doesn't require any special whitespace control. Just define decimal as a terminal, like `/\d+\.\d+/`, which won't allow any whitespace in between. Don't care a rule like `INT "." INT` and it should be fine.

2. Notice that rules are lowercase and terminals are uppercase. It needs to be `DECIMAL_DEFAULT` to work.

**Reply by erezsh** [reply]
Oh btw, your decimal is badly defined. You're using `.` instead of `\.` which matches any character (including a digit) instead of just a dot.

**Reply by Pete-mathpresso** [reply]
Oh, It is my mistake to show . instead of \\. in the code. In fact, it was the same when I used \\. , so I changed it to . , but I showed it without knowing it. lalr parser tokenizes 3.5 to 3 and writes errors lark.exceptions.UnexpectedCharacters: No terminal defined for '.' at line 1 col 2

Surprisingly, earley parser works well for the same grammar. Perhaps there is a mathematical weakness in Lalr's own logic.

The reason for specifying variables as lower case instead of uppercase, in other words, the reason I designated the variables as rules instead of terminals is that I thought only the tree structure could be changed when transforming the parsed tree.

Q. When parsing, the rule is converted to Tree structure and Terminal is converted to Token, but I wonder if there is a way to transform Terminal as i want.

**Reply by Pete-mathpresso** [reply]
Using DECIMAL_DEFAULT instead of decimal_default seems work! surprise.
Then, Could I transform token to what I want? (like float)

**Reply by erezsh** [reply]
Yes, you can transform tokens.
